### DISCUSSION TITLE
envoy变更后无法应用到主仓库

### DISCUSSION THREAD

**Discussion by bergen4** [alibaba/higress#1405] [Q&A]
按照这个文档fork，修改envoy后提交到envoy-1.27分支，但是主项目中子模块的分支是detached状态，无法更新，如果在子模块中强行切换到envoy-1.27分支，能把子模块更新到最新版本，但是会报"正在丢弃xx个提交"，所以请教下我应该如何把最新的envoy变更拉下来，而且不丢弃提交呢？
![QQ_1729215998318](https://github.com/user-attachments/assets/e7277cf8-bb8a-4138-bb7c-1fc48c7e9ca8)

分支detached
![QQ_1729216022153](https://github.com/user-attachments/assets/c8335780-4132-4e1a-8754-00aa3180f250)

**Comment by CH3CHO** [comment]
在子模块里创建一个新分支，然后提交改动，再到主仓库里提交对子模块的改动即可。

**Comment by bergen4** [comment]
"再到主仓库里提交对子模块的改动"，这里首先得拿到子模块的改动，但是现在的问题是在主仓库里面获取子模块的改动会有上面提到的问题。网上找的一些子模块工作模式貌似并不奏效，跟higress的模式好像不太一样
https://git-scm.com/book/zh/v2/Git-%E5%B7%A5%E5%85%B7-%E5%AD%90%E6%A8%A1%E5%9D%97
![QQ_1729217609105](https://github.com/user-attachments/assets/a743986e-51c5-4827-bdf0-1d1a93dfa9dc)

**Comment by CH3CHO** [comment]
> 在子模块里创建一个新分支，然后提交改动，再到主仓库里提交对子模块的改动即可。

你先执行 `make prebuild` 把子模块拉下面，然后做我上面说的事情就行了。

**Comment by bergen4** [comment]
> > 在子模块里创建一个新分支，然后提交改动，再到主仓库里提交对子模块的改动即可。
> 
> 你先执行 `make prebuild` 把子模块拉下面，然后做我上面说的事情就行了。

我是执行了make prebuild的，子模块也拉下来了，envoy子模块处于detach状态。接下来按你的操作，"再到主仓库里提交对子模块的改动"这一步操作不了，会有上面我提到的问题。或者说"提交子模块的改动"姿势不正确吗

**Comment by CH3CHO** [comment]
> "再到主仓库里提交对子模块的改动"这一步操作不了，会有上面我提到的问题。

首先，确认子模块已经没有未提交的改动。你说的“"再到主仓库里提交对子模块的改动"这一步操作不了”是什么意思呢？

**Comment by bergen4** [comment]
> 但是主项目中子模块的分支是detached状态，无法更新，如果在子模块中强行切换到envoy-1.27分支，能把子模块更新到最新版本，但是会报"正在丢弃xx个提交

"再到主仓库里提交对子模块的改动"这一步操作不了，是指：
主仓库中子模块的分支是detached状态，无法更新，执行git pull会报错。如果在主仓库的子模块中强行切换到envoy-1.27分支，虽然能把子模块的更新拉下来，但是会报"正在丢弃xx个提交"的错误
![QQ_1729219645177](https://github.com/user-attachments/assets/69245f64-cafd-44c5-8a77-845e1f68afda)

说明：我是按文档的这种方式新建了envoy子仓库，并在这个子仓库提交变更到分支，然后在主仓库的子模块下去git pull
![QQ_1729219809862](https://github.com/user-attachments/assets/6c2a7844-a075-442d-9417-93b87c872eca)

**Comment by CH3CHO** [comment]
所以你的问题是无法更新子仓库，对吧？

1. detached 状态下是不能 pull 的；
2. 你是因为在 detached 状态下做了提交，所以强制切换分支时被提示会丢弃提交吗？这样的话你只需要在当前 detached 状态下新建一个分支，这样就可以解除 detached 状态了。

你先试一下吧。

**Comment by bergen4** [comment]
我的问题的确是无法更新子仓库。但是有几个疑问
1、为什么执行prebuild后，envoy子仓库的分支就是detached状态了，我没做任何其他改动，这是正常的吗
2、detached状态下是没法更新子仓库代码的，所以我一直不理解你说的"再到主仓库里提交对子模块的改动"，这一步如何操作

另外，回答你提到的疑问"你是因为在 detached 状态下做了提交，所以强制切换分支时被提示会丢弃提交吗？"
我没有在detached状态下做任何提交，主仓库的envoy子仓库执行prebuild后本身就处于envoy-1.27分支，但是处于detached状态没法pull拉取这个分支最新的改动，网上有方法说执行check out envoy-1.27强行切换一下，确实能拉到最新的提交了，但是会有"您正在丢弃xxx个提交"的告警出来。我很疑惑你们是怎么操作的，因为昨天你们有更新envoy和istio的构建问题

**Comment by johnlanni** [comment]
你看下make prebuild的处理脚本，里面有一行 git submodule update --init

**Comment by bergen4** [comment]
> 你看下make prebuild的处理脚本，里面有一行 git submodule update --init

这些脚本我大体都看过了，git submodule update --init初始化子模块然后clone下来，执行完后envoy子模块就是detached状态。执行git submodule update --remote什么的也都不行，昨天我提envoy的构建问题的时候也提了这个问题，你们没问题么
![QQ_1729222595857](https://github.com/user-attachments/assets/d98ab122-5d63-4d75-af79-84e7f5093323)

**Comment by bergen4** [comment]
你们昨天修改higress-group的envoy子仓库后，是怎么更新到higress主仓库去的呢
![QQ_1729222838056](https://github.com/user-attachments/assets/98ee27bd-1bda-4bc2-b44c-67efa341c9ce)

**Comment by johnlanni** [comment]
git submodule update --init 会拉取当前已经commit到主仓库的submodule的。
你如果修改了submodule里的仓库路径，比如改到了你自己fork的分支，那么修改后需要执行 git submodule update --remote，然后将当前对submodule的修改commit 提交，之后再执行 make prebuild。

**Comment by CH3CHO** [comment]
> 你们昨天修改higress-group的envoy子仓库后，是怎么更新到higress主仓库去的呢 ![QQ_1729222838056](https://private-user-images.githubusercontent.com/10844865/377711932-98ee27bd-1bda-4bc2-b44c-67efa341c9ce.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjkyMjMxODcsIm5iZiI6MTcyOTIyMjg4NywicGF0aCI6Ii8xMDg0NDg2NS8zNzc3MTE5MzItOThlZTI3YmQtMWJkYS00YmMyLWI0NGMtNjdlZmEzNDFjOWNlLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDEwMTglMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQxMDE4VDAzNDEyN1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTc3MzMwN2EyYWQyZmJkZWZhOGJlNzE2NWU4MWQwY2EyNjNjMTE4NTNhZDhhN2U5NDIzNzYwNDMwNDFkZDMxMjMmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.v0B-88G_Ed8WGCnyFSCIMI3OAfbqpUusDE7n4wseVcw)

这个就是在主仓库里修改子模块的当前commit，然后提交主仓库里的子模块变更。

**Comment by bergen4** [comment]
假如submodule init之后子仓库的detached状态是正常的，那么把子仓库的变更拉到主仓库并提交的正确操作方式是怎样的呢

**Comment by CH3CHO** [comment]
> 假如submodule init之后子仓库的detached状态是正常的，那么把子仓库的变更拉到主仓库并提交的正确操作方式是怎样的呢

你修改子模块之后，去主仓库里git status看一下。

**Comment by CH3CHO** [comment]
这些都是标准的git操作，不是higress特有的。。。

**Comment by bergen4** [comment]
感谢。更新代码应该使用update --remote
